### PR TITLE
Chemistry rework

### PR DESCRIPTION
**Pull request by ITCups**
## About The Pull Request
Chemistry rework and new additions
Implementations of this doc https://docs.google.com/document/d/13oB79kAXRF1Zi_T8DOcVhy8bM-UbzHgWX-XDfwEFp68/edit#heading=h.f3frw9p5hv9m

## Why It's Good For The Game
Because old chemistry is boring

## Changelog
:cl:
add: Electrolyser
add: Mortar and pestle
add: Centrifuge
add: Stims
add: Nanobots
add: New types of meat
add: Many new chemicals
add: Organ regeneration
add: Blood regeneration
add: Blood circulation value that affects chem metabolism in blood
add: new NSA system
add: Shouting text is a little bit bigger than regular
add: Probably something else, i forgot
add: Added missing recipes for chems
add: Industrial dispenser
add: spilling reagents now creates puddle or powder pile
add: B.I.D.O.Ns
add: new grinding and degradation options
tweak: tweaked hell lot of a things in chemistry
tweak: chemmaster interface
fix: fixed a few other things
refactor: A lot of chemistry code was refactored
/:cl:

### CODE REVIEW

**Comment by Kurgis** [comment]
Can you explain what these are in some fashion? You can just copy+paste the descriptions if they were written somewhere else into the changelog.

**Comment by ITCups** [comment]
its an implementation of this doc https://docs.google.com/document/d/13oB79kAXRF1Zi_T8DOcVhy8bM-UbzHgWX-XDfwEFp68/edit#heading=h.f3frw9p5hv9m
note that some minor features are missing and there are some difference

**Comment by infraredbaron** [comment]
Tested it a lot, no clear issues left. Should be ready for merging, and should be completely ready after adding sources for nanites, new loot to maints, crafting recipes and finally adding machines to chemistry department itself.

File: code/modules/mob/living/carbon/human/human_organs.dm [hunk 1]
```diff
@@ -177,3 +177,87 @@
...
+		if(istype(C))
+			if(!C.is_usable())
+				return TRUE
+	else if(should_have_organ(BP_HEART))
+		var/obj/item/organ/internal/heart/heart = internal_organs_by_name[BP_HEART]
+		if(!istype(heart) || !heart.is_working())
+			return TRUE
+	return FALSE
+
+ge_this_tick
```
Comment (mattroks101) [inline code/modules/mob/living/carbon/human/human_organs.dm]: That doesn't look like that's supposed to be there.

File: code/modules/mob/living/carbon/human/human_powers.dm [hunk 2]
```diff
@@ -192,8 +192,10 @@
 		var/mob/living/carbon/human/H = M
 		if(H.species.name == src.species.name)
 			return
-		to_chat(H, "\red Your nose begins to bleed...")
-		H.drip(1)
+
+		H << "\red Your nose begins to bleed..."
```
Comment (mattroks101) [inline code/modules/mob/living/carbon/human/human_powers.dm]: We need to phase out the \reds wherever possible. Also this needs to be a to_chat().

File: code/modules/organs/blood.dm [hunk 3]
```diff
@@ -7,7 +7,8 @@ var/const/BLOOD_VOLUME_OKAY =    75
 var/const/BLOOD_VOLUME_BAD =     60
 var/const/BLOOD_VOLUME_SURVIVE = 40
 
-/mob/living/carbon/human/var/datum/reagents/vessel // Container for blood and BLOOD ONLY. Do not transfer other chems here.
+/mob/living/carbon
+	var/datum/reagents/vessel // Container for blood and BLOOD ONLY. Do not transfer other chems here.
 /mob/living/carbon/human/var/var/pale = 0          // Should affect how mob sprite is drawn, but currently doesn't.
```
Comment (mattroks101) [inline code/modules/organs/blood.dm]: You edited the line above but this one has the exact same syntax. Why the change?

File: code/modules/reagents/machinery/chem_master.dm [hunk 4]
```diff
@@ -10,29 +10,25 @@
...
 
 /obj/machinery/chem_master/RefreshParts()
-	if(!reagents)
+	/*if(!reagents)
 		create_reagents(10)
 	reagents.maximum_volume = 0
 	for(var/obj/item/weapon/reagent_containers/glass/G in component_parts)
 		reagents.maximum_volume += G.volume
-		G.reagents.trans_to_holder(reagents, G.volume)
+		G.reagents.trans_to_holder(reagents, G.volume)*/
```
Comment (mattroks101) [inline code/modules/reagents/machinery/chem_master.dm]: Is this going to be re-implemented later or something? Because you deleted all the other things here instead of commenting them out.

File: code/modules/reagents/machinery/electrolyzer.dm [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+	var/obj/item/weapon/reagent_containers/beaker
+	var/obj/item/weapon/reagent_containers/separationBeaker
+
+//obj/item/device/makeshiftElectrolyser/update_icon()
+//	if(on)
+	/*if(beaker)
+		icon_state = "mixer1b"
+	else
+		icon_state = "mixer0b"
+	*/
```
Comment (mattroks101) [inline code/modules/reagents/machinery/electrolyzer.dm]: Why is this commented out?

File: code/modules/reagents/reagents.dm [hunk 6]
```diff
@@ -49,9 +49,52 @@
 	var/heating_sound = 'sound/effects/bubbles.ogg'
 	var/list/heating_products
 
+	var/constantMetabolism = FALSE	// if metabolism factor should not change with volume or blood circulation
+
+	var/NSA = 10 // Nerve system accumulations
```
Comment (mattroks101) [inline code/modules/reagents/reagents.dm]: All caps var names look like defines, it would probably be better if this was lowercased, or snake cased somehow.

File: code/modules/reagents/reagents.dm [hunk 7]
```diff
@@ -49,9 +49,52 @@
...
+	if(location == CHEM_BLOOD)
+		if(!constantMetabolism)
+			if(overdose)
+				removed = CLAMP(metabolism * volume/(overdose/2) * M.get_blood_circulation()/100, metabolism * REAGENTS_MIN_EFFECT_MULTIPLIER, metabolism * REAGENTS_MAX_EFFECT_MULTIPLIER)
+			else
+				removed = CLAMP(metabolism * volume/(REAGENTS_OVERDOSE/2) * M.get_blood_circulation()/100, metabolism * REAGENTS_MIN_EFFECT_MULTIPLIER, metabolism * REAGENTS_MAX_EFFECT_MULTIPLIER)
+	removed = round(removed, 0.01)
+	removed = min(removed, volume)
+	if(M.name == "Jhony Mccune")
+		world << "removed [name] = [removed]"
```
Comment (mattroks101) [inline code/modules/reagents/reagents.dm]: This appears to be a debug thing.

File: code/modules/reagents/reagents.dm [hunk 8]
```diff
@@ -49,9 +49,52 @@
...
+	return removed
+
+// "Removed" to multiplier
+// will return multiplier of how much value is more or less than default metabolism amount
+// all chem effects should be multiplied by return of this proc
+/datum/reagent/proc/RTM(var/removed, var/location)
```
Comment (mattroks101) [inline code/modules/reagents/reagents.dm]: All caps proc names are also a little weird. I don't know what the precedent on this is though so if the other maintainers are cool with all caps proc and var names I'm not gonna fight them on it.

File: code/modules/reagents/machinery/electrolyzer.dm [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+		return FALSE
+
+	//check only one reagent present
+	if(!primaryBeaker.reagents.reagent_list.len > 1)
+		return FALSE
+
+	var/datum/chemical_reaction/originalReaction
+	var/activeReagent
+	for(var/datum/reagent/R in primaryBeaker.reagents.reagent_list)
+		for(var/id in GLOB.chemical_reactions_list_by_result)
```
Comment (Carbonhell) [inline code/modules/reagents/machinery/electrolyzer.dm]: the loop isn't needed, just access it directly and check if recipeList isn't null

File: code/modules/reagents/machinery/centrifuge.dm [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+	if(..())
+		return
+
+
+	if(href_list["ejectBeaker"] && !on)
+		if(href_list["ejectBeaker"] == "0")
+			mainBeaker.forceMove(get_turf(src))
+			mainBeaker = null
+		else
+			var/slot = text2num(href_list["ejectBeaker"])
```
[the remark this answers is not in the document]
Comment (Carbonhell) [reply code/modules/reagents/machinery/centrifuge.dm]: Didn't notice one's a machinery and the other's a device (why is a device inside machinery's folder??)
how about using components if they were ported, or instead, use a datum that encapsulates the behaviour and assign it to both the machinery and the device, the datum being the real centrifuge

File: code/modules/reagents/machinery/electrolyzer.dm [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+		if(separationBeaker)
+			separationBeaker.forceMove(get_turf(src))
+			separationBeaker = null
+		update_icon()
+
+	return 1 // update UIs attached to this object
+
+
+
+/obj/item/device/makeshiftElectrolyser
```
Comment (Carbonhell) [inline code/modules/reagents/machinery/electrolyzer.dm]: Same stuff that was said for the centrifuge

File: code/modules/reagents/machinery/electrolyzer.dm [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+
+//obj/item/device/makeshiftElectrolyser/update_icon()
+//	if(on)
+	/*if(beaker)
+		icon_state = "mixer1b"
+	else
+		icon_state = "mixer0b"
+	*/
+
+/obj/item/device/makeshiftElectrolyser/Initialize()
```
Comment (Carbonhell) [inline code/modules/reagents/machinery/electrolyzer.dm]: pointless redefine

File: code/modules/organs/blood.dm (hunk 3 above)
Comment (ITCups) [reply code/modules/organs/blood.dm]: im not really sure this is belong here instead of carbon/human define, but still this way its more distinct and constant

File: code/modules/reagents/machinery/chem_master.dm (hunk 4 above)
Comment (ITCups) [reply code/modules/reagents/machinery/chem_master.dm]: Yeah, i left it so it can be used in future in some other way.
So this will save the trouble.

File: code/modules/reagents/reagents.dm (hunk 8 above)
Comment (ITCups) [reply code/modules/reagents/reagents.dm]: i've changed other occurrences but intend to leave this one, since in feels like defines in some way.

File: code/modules/reagents/machinery/electrolyzer.dm [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+	var/convertion_coefficient = 2
+	var/on = FALSE
+
+/obj/item/weapon/circuitboard/electrolyzer
+	name = T_BOARD("electrolyzer")
+	build_path = /obj/machinery/electrolyzer
+	origin_tech = list(TECH_BIO = 3)
+
+// returns FALSE on errors TRUE on success and -1 if nothing to do
+/proc/electrolysis(var/obj/item/weapon/reagent_containers/primaryBeaker, var/obj/item/weapon/reagent_containers/secondaryBeaker, var/amount)
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/machinery/electrolyzer.dm]: there is two electrolyzers and they have different parent classes.

File: code/game/objects/items.dm [hunk 14]
```diff
@@ -163,7 +163,7 @@
 		R.activate_module(src)
 //		R.hud_used.update_robot_modules_display()
 
-/obj/item/proc/talk_into(mob/M, text)
+/obj/item/proc/talk_into(mob/living/M as mob, message, channel, var/verb = "says", var/datum/language/speaking = null, var/speechVolume)
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/game/objects/items.dm]: `Also proc arguments do not need the var/ prefix.` Yes i know but i will type them anyway, its much easier to search through code if vars definition are constant
also a lot of code is just copypasted and therefore is weird like that

File: code/game/objects/landmarks/debug.dm [hunk 15]
```diff
@@ -0,0 +1,21 @@
+/obj/landmark/debug
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/game/objects/landmarks/debug.dm]: to spawn items from debug menu ingame
very handy, much help.
Not intend to remove

File: code/modules/mob/living/carbon/human/human_damage.dm [hunk 16]
```diff
@@ -212,17 +212,65 @@
 	return ..()
 
 /mob/living/carbon/human/adjustToxLoss(var/amount)
-	if(species.flags & NO_POISON)
+	/*if(species.flags & NO_POISON)
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/mob/living/carbon/human/human_damage.dm]: never intended to leave it, sometimes i forgot to delete code in the process.
Good thing that you are here to point that out.

File: code/modules/mob/living/carbon/human/human_damage.dm [hunk 17]
```diff
@@ -212,17 +212,65 @@
...
+
+	var/heal = amount < 0
+	amount = abs(amount)
+
+	if (!heal)
+		amount = amount * species.toxins_mod
+		if (CE_ANTITOX in chem_effects)
+			amount *= 1 - (chem_effects[CE_ANTITOX] * 0.25)
+
+	var/list/pick_organs = shuffle(internal_organs.Copy())
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/mob/living/carbon/human/human_damage.dm]: This is port from bay, not my code, but since it works i left it as it is. Reworking it was not in my priority list.

File: code/modules/mob/living/living.dm [hunk 18]
```diff
@@ -794,3 +794,17 @@ default behaviour is:
...
 			A.client.images |= static_overlay
+
+/mob/living/proc/vomit()
+	return
+
+/mob/living/proc/adjustNutrition(var/amount)
+	nutrition += amount
+	nutrition = max(0,min(nutrition, max_nutrition))	//clamp the value
+
+/mob/living/proc/is_asystole()
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/mob/living/living.dm]: yes

File: code/modules/organs/external/_external.dm [hunk 19]
```diff
@@ -1023,3 +1017,12 @@ Note that amputating the affected organ does in fact remove the infection from t
 					flavor_text += "a ton of [wound]\s"
 		return english_list(flavor_text)
 
+/obj/item/organ/external/is_usable()
+	//return ..() && !is_stump() && !(status & ORGAN_TENDON_CUT) && (!can_feel_pain() || get_pain() < pain_disability_threshold) && brute_ratio < 1 && burn_ratio < 1
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/organs/external/_external.dm]: Ah, yes, you probably should know about it since you are going to port brainmed. This is how this line is supposed to look with bay12's medicine. But i left it because i forgot to delete it.

File: code/modules/reagents/machinery/centrifuge.dm [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+		on = !on
+		lastActivation = world.time
+		update_icon()
+
+	if(href_list["ejectBeaker"] && !on)
+		if(href_list["ejectBeaker"] == "0")
+			mainBeaker.forceMove(get_turf(src))
+			mainBeaker = null
+		else
+			var/slot = text2num(href_list["ejectBeaker"])
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/machinery/centrifuge.dm]: not needed, its tied to other part of the code and if someone fucks that up they will have hard time debugging it. This way it will runtime and point out the problem.

File: code/modules/reagents/machinery/centrifuge.dm [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+			separationBeakers[slot].forceMove(get_turf(src))
+			separationBeakers.Remove(separationBeakers[slot])
+
+	
+	if(href_list["setTime"] && !on)
+		workTime = text2num(href_list["setTime"]) SECONDS
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/machinery/centrifuge.dm]: also not needed since those values are predefined in template.

File: code/modules/reagents/machinery/centrifuge.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/machinery/centrifuge.dm]: you are talking about whole new system to implement. Yes, it will be much better but its not in scope of this PR.

File: code/modules/mob/living/carbon/carbon.dm [hunk 22]
```diff
@@ -11,11 +11,44 @@
 	..()
 
 	handle_viruses()
-
+	handleNSA()
 	// Increase germ_level regularly
 	if(germ_level < GERM_LEVEL_AMBIENT && prob(30))	//if you're just standing there, you shouldn't get more germs beyond an ambient level
 		germ_level++
 
+/mob/living/carbon/proc/adjustNSA(value, tag)
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/mob/living/carbon/carbon.dm]: i don't agree with this case, too long. Same goes for some cases below

File: code/modules/mob/living/carbon/human/whisper.dm [hunk 23]
```diff
@@ -129,7 +129,7 @@
 	for (var/obj/O in view(message_range, src))
 		spawn (0)
 			if (O)
-				O.hear_talk(src, message, verb, speaking)
+				O.hear_talk(src, message, verb, speaking, 1)
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/mob/living/carbon/human/whisper.dm]: no

File: code/modules/organs/internal/machine.dm [hunk 24]
```diff
@@ -7,14 +7,94 @@
...
 
-/obj/item/organ/cell/replaced()
+/obj/item/organ/internal/cell/Initialize(mapload, ...)
+	. = ..()
+	if(ispath(cell))
+		cell = new cell(src)
+
+/obj/item/organ/internal/cell/proc/percent()
+	if(!cell)
+		return 0
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/organs/internal/machine.dm]: no

File: code/modules/reagents/machinery/centrifuge.dm [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+		if(href_list["ejectBeaker"] == "0")
+			mainBeaker.forceMove(get_turf(src))
+			mainBeaker = null
+		else
+			var/slot = text2num(href_list["ejectBeaker"])
+			separationBeakers[slot].forceMove(get_turf(src))
+			separationBeakers.Remove(separationBeakers[slot])
+
+
+	return 1 // update UIs attached to this object
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/machinery/centrifuge.dm]: no they have certain defines, but they are too long and when no special handling is needed i prefer to use 1 or 0
`// NanoUI flags
#define STATUS_INTERACTIVE 2 // GREEN Visability
#define STATUS_UPDATE 1 // ORANGE Visability
#define STATUS_DISABLED 0 // RED Visability
#define STATUS_CLOSE -1 // Close the interface`

File: code/modules/reagents/reagent_containers/glass/beaker.dm [hunk 26]
```diff
@@ -9,7 +9,7 @@
 /obj/item/weapon/reagent_containers/glass/beaker/Initialize()
 	. = ..()
 	desc += " Can hold up to [volume] units."
-
+
```
[the remark this answers is not in the document]
Comment (ITCups) [reply code/modules/reagents/reagent_containers/glass/beaker.dm]: this is too big of PR you cant expect me to fix every nitpick

File: code/modules/mob/living/carbon/carbon.dm (hunk 22 above)
[the remark this answers is not in the document]
Comment (SpadesSlick) [reply code/modules/mob/living/carbon/carbon.dm]: Then make it "adjust_nsa".

File: code/modules/organs/external/_external.dm (hunk 19 above)
[the remark this answers is not in the document]
Comment (mattroks101) [reply code/modules/organs/external/_external.dm]: >still didn't delete it

File: code/modules/reagents/machinery/electrolyzer.dm (hunk 11 above)
Comment (mattroks101) [reply code/modules/reagents/machinery/electrolyzer.dm]: `makeshiftElectrolyser` is still not properly formatted.
